8313674: (fc) java/nio/channels/FileChannel/BlockDeviceSize.java should test for more block devices

txominpelu opened the pull request:
<!--
Replace this text with a description of your pull request (also remove the surrounding HTML comment markers).
If in doubt, feel free to delete everything in this edit box first, the bot will restore the progress section as needed.
-->

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8313674](https://bugs.openjdk.org/browse/JDK-8313674): (fc) java/nio/channels/FileChannel/BlockDeviceSize.java should test for more block devices (**Enhancement** - P4)


### Reviewers
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**) ⚠️ Review applies to [39e348b3](https://git.openjdk.org/jdk/pull/19021/files/39e348b334c17365c95b14419f56aa605defd20d)
 * [Brian Burkhalter](https://openjdk.org/census#bpb) (@bplb - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19021/head:pull/19021` \
`$ git checkout pull/19021`

Update a local copy of the PR: \
`$ git checkout pull/19021` \
`$ git pull https://git.openjdk.org/jdk.git pull/19021/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19021`

View PR using the GUI difftool: \
`$ git pr show -t 19021`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19021.diff">https://git.openjdk.org/jdk/pull/19021.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19021#issuecomment-2085338045)

AlanBateman commented on test/jdk/java/nio/channels/FileChannel/BlockDeviceSize.java:
```diff
@@ -34,31 +34,40 @@
-        try (FileChannel ch = FileChannel.open(BLK_PATH, READ);
-             RandomAccessFile file = new RandomAccessFile(BLK_FNAME, "r")) {
+        for (String blkFname: BLK_FNAMES) {
+            Path blkPath = getBlkPath(blkFname);
```
You can replace this with Path.of(blkFname), no need for the getBlkPath method.

txominpelu replied:
Indeed, it seems unnecessary to create a method only for that. I've included your suggestion with 08744f4262c7b63b25c7c4e65c430948447f711f.

AlanBateman commented on test/jdk/java/nio/channels/FileChannel/BlockDeviceSize.java:
```diff
@@ -34,31 +34,40 @@
+
+                long size1 = ch.size();
+                long size2 = file.length();
+                if (size1 != size2) {
```
A passing comment on this test is that it just checks that FC::size and RAF::size return the same size. We are sitting on prototype changes that reimplement RAF on FC so it won't be comparing two different implementations if that goes ahead. It will at least test that the size method doesn't throw.

txominpelu replied:
Ok, so if I'm understanding well your comment, it will still make sense to keep this test even if it's only as a validation that the `.size` call won't fail, while the `size1 != size2` will become irrelevant when the changes that you're mentioning get merged.

AlanBateman replied:
That's right, the best we can do then is to check that size is >= 0.

txominpelu replied:
Perfect, I'll leave the check there, and we can change it to `size() >= 0` when the reimplementation of RAF on FC is done.

AlanBateman replied:
The summary in the test description is "Block devices should not report size=0 on Linux". This test will actually pass if the block device size is reported as 0. So I think the test summary needs to change or the test checks that the size is > 0.

txominpelu replied:
You're correct. I believe that until the prototype changes are merged we can keep the equality tests and add a `!= 0` test.

See: https://github.com/openjdk/jdk/pull/19021/commits/e7fd93af9815d1643122d99135df3ac75e6a939c

AlanBateman reviewed (commented):
Extending the test looks okay. If the person assigned is okay with you taking it then it should be okay to continue.

bplb commented on test/jdk/java/nio/channels/FileChannel/BlockDeviceSize.java:
```diff
@@ -34,31 +34,40 @@
 public class BlockDeviceSize {
-    private static final String BLK_FNAME = "/dev/sda1";
-    private static final Path BLK_PATH = Paths.get(BLK_FNAME);
+    private static final List<String> BLK_FNAMES = List.of("/dev/sda1", "/dev/nvme0n1") ;
```
Is this list sufficient? Has any kind of survey been performed to determine whether anything else should be in the list?

txominpelu replied:
The devices included are only the two most popular block devices in Linux. I'd tend to say that having those two should provide enough coverage, but I'm open to including others. For reference, this [link](https://github.com/brianmcgillion/udev/blob/master/rules/rules.d/60-persistent-storage.rules) gives an idea of the rules that udev uses to name devices and hints for example at the fact that I've not included `/dev/mmcblk0` that is popular, for example in the raspberry pis.

bplb replied:
I'm not familiar enough with which devices need to be covered. If there are more devices of non-negligible interest, however, and adding them would neither destabilize the test nor substantially increase the test's running time, then I would not hesitate to add a few. On the other if you think this is enough, I cannot disagree.

txominpelu replied:
I've thought about it and in the end I've also added `/dev/xvda1`. That should allow running the test in virtualized environments, what should increase significantly the coverage. I believe those three devices should be a very representative set of the environments where this test will be run.

txominpelu commented on test/jdk/java/nio/channels/FileChannel/BlockDeviceSize.java:
```diff
@@ -40,15 +40,11 @@
     public static void main(String[] args) throws Throwable {
         for (String blkFname: BLK_FNAMES) {
-            Path blkPath = getBlkPath(blkFname);
+            Path blkPath = Paths.get(blkFname);
```
Thanks ! I've done that in https://github.com/openjdk/jdk/pull/19021/commits/39e348b334c17365c95b14419f56aa605defd20d

txominpelu commented:
/integrate

txominpelu commented:
Oops sorry, I thought that integrating meant merging the latest changes from master into my PR but I misunderstood and that's just a side effect of calling `/integrate`, now I get that integrate is meant to merge the PR and as a part of that merge it first merges the latest changes from master.

AlanBateman reviewed (approved):
The update to the test looks okay to me, and I think useful to have the test changed to check > 0 is good. It might be clearer to have it fail if <= 0 of course.

I assume you'll bump the end date in the copyright header.

txominpelu commented:
> The update to the test looks okay to me, and I think useful to have the test changed to check > 0 is good. It might be clearer to have it fail if <= 0 of course.
> 
> I assume you'll bump the end date in the copyright header.



>

Thanks !

I've updated the copyright in: https://github.com/openjdk/jdk/pull/19021/commits/78af373b3df429ea2683293517f2c75b0c12b00e

And updated the failure condition to `<=` in: https://github.com/openjdk/jdk/pull/19021/commits/868a40fddb6d6167e00cc0dbfc39470c9cbf4985

bplb commented:
@txominpelu You need to remove the integration blocker by changing the title of this PR to match the summary of the issue
```
8313674: (fc) java/nio/channels/FileChannel/BlockDeviceSize.java should test for more block devices
```

txominpelu commented:
> @txominpelu You need to remove the integration blocker by changing the title of this PR to match the summary of the issue
> 
> ```
> 8313674: (fc) java/nio/channels/FileChannel/BlockDeviceSize.java should test for more block devices
> ```

Thanks @bplb , I've fixed the title of the PR.

bplb commented:
@txominpelu I think you will need to reissue the `/integrate` command.

txominpelu commented:
/integrate

bplb commented:
/sponsor

AlanBateman commented on test/jdk/java/nio/channels/FileChannel/BlockDeviceSize.java:
```diff
@@ -24,41 +24,47 @@
+                System.err.println("File " + blkFname + " not found." +
+                        " Skipping test");
+            } catch (AccessDeniedException ade) {
+                throw new RuntimeException("Access to " + blkFname + " is denied."
```
Looking at it again, this should be SkipException rather RuntimeException, otherwise the test will fail when denied.
